### PR TITLE
InMemoryMessageStore is not thread save

### PR DESCRIPTION
**Pull request by ingoschaal**
Fixed a check-then-act situation in InMemoryMessageStore at add(final QueuedRequest entity)...
Also added some Junit Test for InMemoryMessageStore.

### CODE REVIEW

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java [hunk 1]
```diff
@@ -39,40 +39,54 @@
     private final ConcurrentMap<String, BlockingQueue<QueuedRequest>> requestQueueMap = new ConcurrentHashMap<>();
 
     @Override
-    public void add(QueuedRequest entity) {
+    public void add(final QueuedRequest entity) {
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: pls, remove all unnecessary final keyword.
(see https://github.com/eclipse/leshan/wiki/Code-&-design-guidelines)

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java [hunk 2]
```diff
@@ -39,40 +39,54 @@
...
+        // retrieve its own instance of
+        // Queue and add the entity to the respective instances. But only one of
+        // them wins the putIfAbsent method.
+        // To not lose messages, this check will prevent.
+        final BlockingQueue<QueuedRequest> allreadyExisting = requestQueueMap.putIfAbsent(endpoint, requestQueue);
+        if (allreadyExisting != null) {
+            allreadyExisting.add(entity);
+        } else {
+            requestQueue.add(entity);
+        }
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: it seems better :)

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java [hunk 3]
```diff
@@ -39,40 +39,54 @@
...
 
     @Override
-    public boolean isEmpty(String endpoint) {
-        return getMessageQueueForEndpoint(endpoint).isEmpty();
+    public boolean isEmpty(final String endpoint) {
+        if (queueExist(endpoint)) {
+            return requestQueueMap.get(endpoint).isEmpty();
+        }
+        return true;
+
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: It seems that we can get NPE if between `if` and `return` the value for `endpoint` is removed from the map.
So I prefer something like : 
```java
  BlockingQueue<QueuedRequest> requests = requestQueueMap.get(endpoint);
  if (requests != null){
      requests.isEmpty();
  }
  return true;
```

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java [hunk 4]
```diff
@@ -39,40 +39,54 @@
...
     @Override
-    public QueuedRequest retrieveFirst(String endpoint) {
+    public QueuedRequest retrieveFirst(final String endpoint) {
         LOG.trace("Retrieve first for endpoint {}", endpoint);
-        BlockingQueue<QueuedRequest> requests = getMessageQueueForEndpoint(endpoint);
-        return requests.peek();
+        if (queueExist(endpoint)) {
+            return requestQueueMap.get(endpoint).peek();
+        }
+        return null;
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: same as above

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java [hunk 5]
```diff
@@ -39,40 +39,54 @@
...
-        }
+        requestQueueMap.remove(endpoint);
     }
 
     @Override
-    public void deleteFirst(String endpoint) {
+    public void deleteFirst(final String endpoint) {
         LOG.debug("Delete first entity of endpoint {}", endpoint);
-        BlockingQueue<QueuedRequest> requests = requestQueueMap.get(endpoint);
+        final BlockingQueue<QueuedRequest> requests = requestQueueMap.get(endpoint);
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: same as above

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java [hunk 6]
```diff
@@ -84,19 +98,26 @@ public void deleteFirst(String endpoint) {
...
+            return Collections.emptyList();
         }
     }
 
-    private BlockingQueue<QueuedRequest> getMessageQueueForEndpoint(String endpoint) {
-        return requestQueueMap.get(endpoint) != null ? requestQueueMap.get(endpoint)
-                : new LinkedBlockingQueue<QueuedRequest>();
+    private boolean queueExist(final String endpoint) {
+        return requestQueueMap.get(endpoint) != null ? true : false;
+    }
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: we can remove this method (if my remarks above make sense)

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java [hunk 7]
```diff
@@ -84,19 +98,26 @@ public void deleteFirst(String endpoint) {
      * @param endpoint client's endpoint
      * @return list of queue request entities in order of processing.
      */
-    public List<QueuedRequest> retrieveAll(String endpoint) {
+    public List<QueuedRequest> retrieveAll(final String endpoint) {
         LOG.debug("Retrieve all for endpoint {}", endpoint);
-        BlockingQueue<QueuedRequest> requests = getMessageQueueForEndpoint(endpoint);
-        if (requests.isEmpty()) {
-            return Collections.emptyList();
+        if (queueExist(endpoint)) {
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: same as above

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java [hunk 8]
```diff
@@ -84,19 +98,26 @@ public void deleteFirst(String endpoint) {
...
+        return requestQueueMap.get(endpoint) != null ? true : false;
+    }
+
+    /**
+     * For test onyl.
+     * 
+     * @return ConcurrentMap<String, BlockingQueue<QueuedRequest>>
+     */
+    protected ConcurrentMap<String, BlockingQueue<QueuedRequest>> getRequestQueueMap() {
+        return this.requestQueueMap;
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: I don't really like the idea to expose the requestQueueMap just for testing.
Maybe we can add a getSize() on this class, with retrieveAll.getSize(), we will cover all your needs.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java (hunk 1 above)
Comment (ingoschaal) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: yes i will do this of course.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java (hunk 3 above)
Comment (ingoschaal) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: Ok, that is  possible, you are right.
My thought was not to have so much duplicate code, therefore the method. I will change this and remove the method.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java (hunk 6 above)
Comment (ingoschaal) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: yes, as mentioned above

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java (hunk 8 above)
Comment (balsmn) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: FYI, I have added a similar `getQueueSize(endpoint)` method and removed `retrieveAll(endpoint)` method from InMemoryMessageStore as part of the PR #158.

Comment (ingoschaal) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/InMemoryMessageStore.java]: OK good point.  Now I added this getQueueSize method, looks good also for testing use.

**Comment by sbernard31** [comment]
Integrated in master commit (ddaa6e4744d35ed8c9afe34a37ef44a85ec76a0f)

**Pull request closed without merging**